rosidl_typesupport_opensplice_c/resource/msg__type_support_c.cpp.em

```diff
@@ -51,57 +50,78 @@ extern "C"
+    elif isinstance(member.type, String):
+        keys.add('rosidl_generator_c/string.h')
+        keys.add('rosidl_generator_c/string_functions.h')
+    elif isinstance(member.type, WString):
+        keys.add('rosidl_generator_c/u16string.h')
+        keys.add('rosidl_generator_c/u16string_functions.h')
     for key in keys:
         if key not in includes:
             includes[key] = set([])
-        includes[key].add(field.name)
+        includes[key].add(member.name)
 }@
 @[for key in sorted(includes.keys())]@
-#include "@(key)"  // @(', '.join(includes[key]))
+@[  if key in include_directives]@
+// already included above
+// @
+@[  else]@
+@{include_directives.add(key)}@
+@[  end if]@
+#include "@(key)"  // @(', '.join(sorted(includes[key])))
 @[end for]@
 
 // forward declare type support functions
 @{
 forward_declares = {}
-for field in spec.fields:
-    if not field.type.is_primitive_type():
-        key = (field.type.pkg_name, field.type.type)
+for member in message.structure.members:
+    _type = member.type
+    if isinstance(_type, NestedType):
+       _type = member.type.basetype
+
+    if isinstance(_type, NamespacedType):
+        key = (', '.join(_type.namespaces), _type.name)
         if key not in includes:
             forward_declares[key] = set([])
-        forward_declares[key].add(field.name)
+        forward_declares[key].add(member.name)
 }@
 @[for key in sorted(forward_declares.keys())]@
-@[  if key[0] != pkg]@
-ROSIDL_TYPESUPPORT_OPENSPLICE_C_IMPORT_@(pkg)
+@[  if key[0] != package_name]@
+ROSIDL_TYPESUPPORT_OPENSPLICE_C_IMPORT_@(package_name)
 @[  end if]@
 const rosidl_message_type_support_t *
-  ROSIDL_TYPESUPPORT_INTERFACE__MESSAGE_SYMBOL_NAME(rosidl_typesupport_opensplice_c, @(key[0]), msg, @(key[1]))();
+  ROSIDL_TYPESUPPORT_INTERFACE__MESSAGE_SYMBOL_NAME(
+  rosidl_typesupport_opensplice_c, @(key[0]), @(key[1]))();
 @[end for]@
 
 @# // Make callback functions specific to this message type.
 @{
-__dds_msg_type_prefix = "{0}::{1}::dds_::{2}_".format(
-  spec.base_type.pkg_name, subfolder, spec.base_type.type)
+__dds_msg_type_prefix = '::'.join(message.structure.type.namespaces + ['dds_'] + [message.structure.type.name]) + '_'
+__ros_msg_type_prefix = '__'.join(message.structure.type.namespaces + [message.structure.type.name])
+__static_suffix = '__'.join(message.structure.type.namespaces + [message.structure.type.name])
 }@
-using __dds_msg_type = @(__dds_msg_type_prefix);
-using __ros_msg_type = @(pkg)__@(subfolder)__@(type);
+using __dds_msg_type_@(__static_suffix) = @(__dds_msg_type_prefix);
+using __ros_msg_type_@(__static_suffix) = @(__ros_msg_type_prefix);
 
-static @(__dds_msg_type_prefix)TypeSupport _type_support;
+static @(__dds_msg_type_prefix)TypeSupport _type_support_@(__static_suffix);
 
 static const char *
-register_type(void * untyped_participant, const char * type_name)
+register_type_@(__static_suffix)(void * untyped_participant, const char * type_name)
```

Comment:
The suffix gives the function a message type specific name. Since a single cpp file can contain the type support for multiple messages (e.g. in the case of services and actions) they need separate names.